tests: `test_profile_name_exists` fails.

```console
assert 'Username already exists.' in '\n<!DOCTYPE html>\n<html lang="en">\n  <head>\n    <meta charset="utf-8" />\n  </head>\n  <body>\n<form method="POST"...file" class="btn btn-primary"><i class="fa fa-check"></i> Update profile</button>\n<
/div>\n</form>\n  </body>\n</html>'                            
E            +  where '\n<!DOCTYPE html>\n<html lang="en">\n  <head>\n    <meta charset="utf-8" />\n  </head>\n  <body>\n<form method="POST"...file" class="btn btn-primary"><i class="fa fa-check"></i> Update profile</button>\n</div>\n</form>\n  </body>\n
</html>' = <bound method BaseResponse.get_data of <<class 'pytest_flask.plugin.JSONResponse'> 1447 bytes [200 OK]>>(as_text=True)
E            +    where <bound method BaseResponse.get_data of <<class 'pytest_flask.plugin.JSONResponse'> 1447 bytes [200 OK]>> = <<class 'pytest_flask.plugin.JSONResponse'> 1447 bytes [200 OK]>.get_data

tests/test_views.py:174: AssertionError
```

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -36,33 +36,22 @@ jobs:
 
           include:
           - db-service: postgresql9
-            DB: postgresql
-            POSTGRESQL_VERSION: POSTGRESQL_9_LATEST
-            SQLALCHEMY_DATABASE_URI: "postgresql+psycopg2://invenio:invenio@localhost:5432/invenio"
+            DB: postgresql9
             EXTRAS: "all,postgresql"
 
           - db-service: postgresql11
-            DB: postgresql
-            POSTGRESQL_VERSION: POSTGRESQL_11_LATEST
-            SQLALCHEMY_DATABASE_URI: "postgresql+psycopg2://invenio:invenio@localhost:5432/invenio"
+            DB: postgresql9
             EXTRAS: "all,postgresql"
 
           - db-service: mysql5
-            DB: mysql
-            MYSQL_VERSION: MYSQL_5_LATEST
-            SQLALCHEMY_DATABASE_URI: "mysql+pymysql://invenio:invenio@localhost:3306/invenio"
+            DB: mysql5
             EXTRAS: "all,mysql"
 
           - db-service: mysql8
-            DB: mysql
-            MYSQL_VERSION: MYSQL_8_LATEST
-            SQLALCHEMY_DATABASE_URI: "mysql+pymysql://invenio:invenio@localhost:3306/invenio"
+            DB: mysql8
             EXTRAS: "all,mysql"
 
     env:
-      SQLALCHEMY_DATABASE_URI: ${{matrix.SQLALCHEMY_DATABASE_URI}}
-      POSTGRESQL_VERSION: ${{matrix.POSTGRESQL_VERSION}}
-      MYSQL_VERSION: ${{matrix.MYSQL_VERSION}}
       DB: ${{ matrix.DB }}
 
     steps:

--- a/invenio_userprofiles/api.py
+++ b/invenio_userprofiles/api.py
@@ -29,13 +29,13 @@ def _get_current_userprofile():
     if current_user.is_anonymous:
         return AnonymousUserProfile()
 
-    profile = g.get(
-        'userprofile',
-        UserProfile.get_by_userid(current_user.get_id()))
+    profile = getattr(
+        g, 'userprofile', UserProfile.get_by_userid(current_user.get_id()))
 
     if profile is None:
         profile = UserProfile(user_id=int(current_user.get_id()))
         g.userprofile = profile
+
     return profile
 
 

--- a/invenio_userprofiles/forms.py
+++ b/invenio_userprofiles/forms.py
@@ -65,12 +65,25 @@ class ProfileForm(FlaskForm):
             raise ValidationError(e)
 
         try:
+            # Check if username is already taken (if the username is *not*
+            # found a NoResultFound exception is raised).
             user_profile = UserProfile.get_by_username(field.data)
-            if current_userprofile.is_anonymous or \
-                    (current_userprofile.user_id != user_profile.user_id and
-                     field.data != current_userprofile.username):
-                # NOTE: Form validation error.
-                raise ValidationError(_('Username already exists.'))
+
+            # NOTE: Form validation error.
+            msg = _('Username already exists.')
+
+            if current_userprofile.is_anonymous:
+                # We are handling a new sign up (i.e. anonymous user) AND a
+                # the username already exists. Fail.
+                raise ValidationError(msg)
+            else:
+                # We are handling a user editing their profile AND a
+                # the username already exists.
+                is_same_user = \
+                    current_userprofile.user_id == user_profile.user_id
+                if not is_same_user:
+                    # Username already taken by another user.
+                    raise ValidationError(msg)
         except NoResultFound:
             return
 

--- a/invenio_userprofiles/forms.py
+++ b/invenio_userprofiles/forms.py
@@ -68,7 +68,6 @@ class ProfileForm(FlaskForm):
             # Check if username is already taken (if the username is *not*
             # found a NoResultFound exception is raised).
             user_profile = UserProfile.get_by_username(field.data)
-
             # NOTE: Form validation error.
             msg = _('Username already exists.')
 
@@ -80,7 +79,7 @@ class ProfileForm(FlaskForm):
                 # We are handling a user editing their profile AND a
                 # the username already exists.
                 is_same_user = \
-                    current_userprofile.user_id == user_profile.user_id
+                    current_user.id == user_profile.user_id
                 if not is_same_user:
                     # Username already taken by another user.
                     raise ValidationError(msg)

--- a/run-tests.sh
+++ b/run-tests.sh
@@ -18,14 +18,13 @@ set -o nounset
 
 # Always bring down docker services
 function cleanup() {
-    docker-services-cli down
+    eval "$(docker-services-cli down --env)"
 }
 trap cleanup EXIT
 
-pydocstyle invenio_userprofiles
 python -m check_manifest --ignore ".*-requirements.txt"
-sphinx-build -qnNW docs docs/_build/html # Fails due to intersphinx invenio-accounts 403
-docker-services-cli up ${DB}
+sphinx-build -qnNW docs docs/_build/html
+eval "$(docker-services-cli up --db ${DB:-postgresql} --env)"
 python -m pytest
 tests_exit_code=$?
 exit "$tests_exit_code"

--- a/tests/test_views.py
+++ b/tests/test_views.py
@@ -145,8 +145,9 @@ def test_profile_view(app):
 def test_profile_name_exists(app):
     """Test the profile view."""
     app.config['USERPROFILES_EMAIL_ENABLED'] = False
+    error_msg = 'Username already exists.'
 
-    with app.test_request_context():
+    with app.app_context():
         profile_url = url_for('invenio_userprofiles.profile')
 
     # Create an existing user
@@ -158,7 +159,7 @@ def test_profile_name_exists(app):
         assert client.get(profile_url).status_code == 200
         resp = client.post(profile_url, data=prefix('profile', dict(
             username='existingname', full_name='Valid Name',)))
-        assert 'has-error' not in resp.get_data(as_text=True)
+        assert error_msg not in resp.get_data(as_text=True)
 
     # Create another user and try setting username to same as above user.
     with app.test_client() as client:
@@ -171,19 +172,44 @@ def test_profile_name_exists(app):
             username='existingname', full_name='Another name',
         )))
         assert resp.status_code == 200
-        assert 'Username already exists.' in resp.get_data(as_text=True)
+        assert error_msg in resp.get_data(as_text=True)
 
-        # Now set it to something else and do it twice.
+
+def test_profile_case_change(app):
+    """Test the profile view."""
+    app.config['USERPROFILES_EMAIL_ENABLED'] = False
+    error_msg = 'Username already exists.'
+
+    with app.app_context():
+        profile_url = url_for('invenio_userprofiles.profile')
+
+    with app.test_client() as client:
+        # Create a user
+        sign_up(app, client)
+        login(app, client)
+        resp = client.get(profile_url)
+        assert resp.status_code == 200
+
         data = prefix('profile', dict(
             username='valid', full_name='Another name', ))
 
+        # Set the name first time
         resp = client.post(profile_url, data=data)
         assert resp.status_code == 200
-        assert 'has-error' not in resp.get_data(as_text=True)
+        assert error_msg not in resp.get_data(as_text=True)
+
+        # Set the name second time
+        resp = client.post(profile_url, data=data)
+        assert resp.status_code == 200
+        assert error_msg not in resp.get_data(as_text=True)
+
+        # Change case of the username
+        data = prefix('profile', dict(
+            username='Valid', full_name='Another name', ))
 
         resp = client.post(profile_url, data=data)
         assert resp.status_code == 200
-        assert 'Username already exists.' in resp.get_data(as_text=True)
+        assert error_msg not in resp.get_data(as_text=True)
 
 
 def test_send_verification_form(app):
@@ -212,6 +238,7 @@ def test_send_verification_form(app):
 def test_change_email(app):
     """Test send verification form."""
     mail = app.extensions['mail']
+    error_msg = 'Username already exists.'
 
     with app.test_request_context():
         profile_url = url_for('invenio_userprofiles.profile')
@@ -237,37 +264,53 @@ def test_change_email(app):
             email_repeat=app.config['TEST_USER_EMAIL'],
         ))
 
-        # Test that current_user stops validation of email
-        client.post(profile_url, data=data)
-        assert resp.status_code == 200
-        assert 'has-error' not in resp.get_data(as_text=True)
-
         # Test existing email of another user.
         data['profile-email_repeat'] = data['profile-email'] = email1
         resp = client.post(profile_url, data=data)
-        assert 'has-error' in resp.get_data(as_text=True)
+        assert 'exiting@example.org is already associated with an account.' \
+            in resp.get_data(as_text=True)
 
         # Test empty email
         data['profile-email_repeat'] = data['profile-email'] = ''
         resp = client.post(profile_url, data=data)
-        assert 'has-error' in resp.get_data(as_text=True)
+        assert "Email not provided" in resp.get_data(as_text=True)
 
         # Test not an email
         data['profile-email_repeat'] = data['profile-email'] = 'sadfsdfs'
         resp = client.post(profile_url, data=data)
-        assert 'has-error' in resp.get_data(as_text=True)
+        assert "Invalid email address" in resp.get_data(as_text=True)
 
         # Test different emails
         data['profile-email_repeat'] = 'typo@example.org'
         data['profile-email'] = 'new@example.org'
         resp = client.post(profile_url, data=data)
-        assert 'has-error' in resp.get_data(as_text=True)
+        assert 'Email addresses do not match.' in resp.get_data(as_text=True)
 
-        # Test whitespace
+
+def test_change_email_whitespace(app):
+    """Test send verification form."""
+    mail = app.extensions['mail']
+
+    with app.test_request_context():
+        profile_url = url_for('invenio_userprofiles.profile')
+
+    with app.test_client() as client:
+        sign_up(app, client)
+        login(app, client)
+        resp = client.get(profile_url)
+        assert resp.status_code == 200
+
+        data = prefix('profile', dict(
+            username='test',
+            full_name='Test User',
+            email=app.config['TEST_USER_EMAIL'],
+            email_repeat=app.config['TEST_USER_EMAIL'],
+        ))
+
         with mail.record_messages() as outbox:
             assert len(outbox) == 0
             data['profile-email_repeat'] = data['profile-email'] = 'new@ex.org'
             resp = client.post(profile_url, data=data)
-            assert 'has-error' not in resp.get_data(as_text=True)
+            assert 'Invalid email address' not in resp.get_data(as_text=True)
             # Email was sent for email address confirmation.
             assert len(outbox) == 1

--- a/tests/test_views.py
+++ b/tests/test_views.py
@@ -183,7 +183,7 @@ def test_profile_name_exists(app):
 
         resp = client.post(profile_url, data=data)
         assert resp.status_code == 200
-        assert 'has-error' not in resp.get_data(as_text=True)
+        assert 'Username already exists.' in resp.get_data(as_text=True)
 
 
 def test_send_verification_form(app):

